src/dataflow/src/sink/kafka.rs

```diff
@@ -818,154 +910,99 @@ where
-                                Ok(()) => {
-                                    // sanity check for the continuous updating
-                                    // of the write frontier below
-
-                                    s.assert_progress(ts);
-                                    progress_update.replace(ts.clone());
-
-                                    s.ready_rows.pop_front();
-                                    SendState::BeginTxn
-                                }
-                                Err(e) => s.transition_on_txn_error(s.send_state, *ts, e),
-                            }
-                        }
-                        SendState::AbortTxn => {
-                            let result = if s.transactional {
-                                s.abort_transaction(s.txn_timeout).await
-                            } else {
-                                Ok(())
-                            };
-
-                            match result {
-                                Ok(()) => SendState::BeginTxn,
-                                Err(e) => s.transition_on_txn_error(s.send_state, *ts, e),
-                            }
-                        }
-                        SendState::Shutdown => {
-                            s.shutdown_flag.store(true, Ordering::SeqCst);
-                            // Indicate that the sink is closed to everyone else who
-                            // might be tracking its write frontier.
-                            info!("shutting down kafka sink: {}", &s.name);
-                            break;
-                        }
+            // Can't use `?` when the return type is a `bool` so wrap up all the tx error handling
+            macro_rules! retry_txn_with_error {
+                (|$s:ident, $producer:ident| $body:block) => {
+                    if let Err(_error) = $s
+                        .retry_on_txn_error(|$producer| async move { $body })
+                        .await
+                    {
+                        return true;
+                    }
+                };
+            }
+
+            if is_active_worker
+                && matches!(s.send_state, SendState::Init)
+                // Removing this check would cause us to start writing out END consistency rows as
+                // soon as the sink starts up.  There isn't anything inherently wrong with this but
+                // we need to update testdrive to be able to check a reasonable condition that this
+                // is correct.  That is, td currently checks we start with a BEGIN and then END.
+                // We'd need to be able to do a search for a BEGIN to properly test were we to
+                // remove this check.  This would fix #9514.
+                && s.ready_rows.front().is_some()
+            {
+                if s.transactional {
+                    retry_txn_with_error!(|s, producer| { producer.init_transactions().await });
```

Comment:
Using the macro above this would become:

```suggestion
                    bail_err!(s.retry_txn_with_error(|producer| producer.init_transactions()).await))
```